linera-service/src/linera-exporter/runloops/validator_exporter.rs

```diff
@@ -0,0 +1,215 @@
+
+use crate::{common::ExporterError, dispatch, storage::ExporterStorage};
+
+struct ExportTask<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    node: GrpcClient,
+    destination_id: DestinationId,
+    storage: &'a ExporterStorage<S>,
+}
+
+pub(crate) struct Exporter<S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    node_provider: Arc<GrpcNodeProvider>,
+    destination_id: DestinationId,
+    storage: ExporterStorage<S>,
+    destination_config: Destination,
+    work_queue_size: usize,
+}
+
+impl<S> Exporter<S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    pub(super) fn new(
+        node_provider: Arc<GrpcNodeProvider>,
+        destination_id: DestinationId,
+        storage: ExporterStorage<S>,
+        destination_config: Destination,
+        work_queue_size: usize,
+    ) -> Self {
+        Self {
+            node_provider,
+            destination_id,
+            storage,
+            destination_config,
+            work_queue_size,
+        }
+    }
+
+    pub(super) async fn run_with_shutdown<F: IntoFuture<Output = ()>>(
+        self,
+        signal: F,
+    ) -> anyhow::Result<()> {
+        let address = self.destination_config.address();
+        let node = self.node_provider.make_node(&address)?;
+
+        let export_task = ExportTask::new(node, self.destination_id, &self.storage);
+        let (mut task_queue, task_receiver) =
+            TaskQueue::new(self.work_queue_size, self.destination_id, &self.storage);
+
+        tokio::select! {
+
+            biased;
+
+            _ = signal => {},
+
+            res = task_queue.run() => res?,
+
+            res = export_task.run(task_receiver) => res?,
+
+        };
+
+        Ok(())
+    }
+}
+
+impl<'a, S> ExportTask<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(
+        node: GrpcClient,
+        destination_id: DestinationId,
+        storage: &'a ExporterStorage<S>,
+    ) -> ExportTask<'a, S> {
+        ExportTask {
+            node,
+            destination_id,
+            storage,
+        }
+    }
+
+    async fn run(
+        &self,
+        mut receiver: Receiver<Arc<ConfirmedBlockCertificate>>,
+    ) -> anyhow::Result<()> {
+        let delivery = CrossChainMessageDelivery::NonBlocking;
+        while let Some(block) = receiver.recv().await {
+            let hash = block.hash();
+            let method = |certificate, delivery| {
+                self.node
+                    .handle_confirmed_certificate(certificate, delivery)
+            };
+
+            // As the linera-rpc client already contains a retry loop, the loop below is mainly for the blobs.
+            loop {
+                match dispatch!(method, log = hash, (*block).clone(), delivery) {
+                    Ok(_) => break,
+
+                    Err(NodeError::BlobsNotFound(blobs)) => self.upload_blobs(blobs).await?,
+
+                    Err(e) => Err(e)?,
+                }
+            }
```

Comment:
We could later also defend against a malicious validator here: If the request blobs we've already sent, or blobs that the block doesn't actually depend on, we should stop.